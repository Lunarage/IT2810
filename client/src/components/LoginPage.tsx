import React, {Component} from 'react';


class LoginPage extends Component<Props, State> {
    render() {
        return (
            <div className={"login-page"} >
                <div className={"login-section"}>

                    <h2 className={"login-page-title"}>Logg inn</h2>
                    <form className={"login-page-form"}>
                        <div className={"form-group username"}>
                            <label className={"input-label"} htmlFor={"username"}>Brukernavn</label>
                            <input name={"username"} className={"input"} autoFocus placeholder={"OlaNormann"}/>
                        </div>

                        <div className={"form-group password"}>
                            <label className={"input-label"} htmlFor={"password"}>Passord</label>
                            <input name={"password"} className={"input"} type={"password"} autoFocus
                                   placeholder={"Passord"}/>
                        </div>

                        <div className={"form-group"}>
                            <button type={"submit"}>Logg inn</button>
                        </div>
                    </form>
                </div>

            </div>

        );
    }


}

export default LoginPage;

interface Props{
    
}

interface State{

}