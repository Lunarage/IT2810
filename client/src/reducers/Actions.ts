import HttpClient from "../modules/HttpClient";
import { SessionStorageWrapper, LocalStorageWrapper } from "../modules/Storage";

export const toggle_loggedIn = (boolVal: boolean) => {
    return {
        type: "TOGGLE_LOGGED_IN",
        payload: boolVal
    } as const
};

export const set_username = ( text: string) => {
    return {
        type: "SET_USERNAME",
        payload: text
    } as const
};

export const logout = (boolVal: boolean) => {
    return {
        type: "LOGOUT",
        payload: boolVal
    } as const 
};
