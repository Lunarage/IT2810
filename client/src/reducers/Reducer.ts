import {
    toggle_loggedIn,
    logout,
    set_username,
    setStoreSearchState,
} from "./Actions";
import { createStore, Store } from "redux";
import { Movie } from "../types/DatabaseTypes";

type Actions =
    | ReturnType<typeof toggle_loggedIn>
    | ReturnType<typeof logout>
    | ReturnType<typeof set_username>
    | ReturnType<typeof setStoreSearchState>;

// The type for holding search information and result
export type StoreSearchState = {
    inputSearch: string;
    inputOrderDir: "ASC" | "DESC";
    inputTitleType: string;
    pageActivePage: number;
    pageTotalPages: number;
    pageDisabled: boolean;
    searchStatus: "none" | "waiting" | "success" | "failure";
    searchErrorMessage: string | null;
    searchMovies: Movie[];
};

//Creates the types AppState
export type AppState = {
    loggedIn: boolean;
    userName?: string | null;
    searchState: StoreSearchState;
};

//Sets initial root state for the app
const initialAppState: AppState = {
    loggedIn: false,
    userName: null,
    searchState: {
        inputSearch: "",
        inputOrderDir: "ASC",
        inputTitleType: "",
        pageActivePage: 1,
        pageTotalPages: 2,
        pageDisabled: false,
        searchStatus: "none",
        searchErrorMessage: null,
        searchMovies: [],
    },
};
//Root reducer for the App, handels the states loggedIn/Loggedout and userName
export const rootReducer = (
    state: AppState = initialAppState,
    action: Actions
) => {
    switch (action.type) {
        case "TOGGLE_LOGGED_IN":
            return {
                ...state,
                loggedIn: action.payload,
            };

        case "SET_USERNAME":
            return {
                ...state,
                userName: action.payload,
            };

        case "LOGOUT":
            return {
                ...state,
                loggedIn: action.payload,
                userName: null,
            };
        case "SET_SEARCH_STATE":
            return {
                ...state,
                searchState: action.payload,
            };
        default:
            return state;
    }
};
//Configure store
function configureStore(): Store<AppState> {
    return createStore(
        rootReducer,
        (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
            (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    );
}

//Create and exports store
export const store = configureStore();
