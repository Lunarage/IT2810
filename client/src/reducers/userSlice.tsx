import { createSlice } from "@reduxjs/toolkit";
import HttpClient from "../modules/HttpClient";
import { SessionStorageWrapper, LocalStorageWrapper } from "../modules/Storage";

// redux komponent som lager en "slice" av store hvor vi kan definere en reducer og actions som handler om en bruker logget inn eller ikke
const loginSlice = createSlice({
    name: "login/logout",
    initialState: { value: false, username: null },

    reducers: {
        logIn: (state, action) => {
            const baseURL = "http://it2810-22.idi.ntnu.no:3000";
            const client = new HttpClient(baseURL);
            const localStorage = new LocalStorageWrapper();
            localStorage.set("username", action.payload);
            client.createUser(action.payload);
            state.value = true;
            state.username = action.payload;
        },
        logOut: (state) => {
            const localStorage = new LocalStorageWrapper();
            localStorage.remove("username");
            state.value = false;
            state.username = null;
        },
    },
});

export const { logIn, logOut } = loginSlice.actions;

export default loginSlice.reducer;

/* These sites are used as src for this code: 
https://codesandbox.io/s/github/reduxjs/redux-essentials-counter-example/tree/master/?from-embed=&file=/src/features/counter/counterSlice.js:49-83
https://redux.js.org/tutorials/essentials/part-2-app-structure */
