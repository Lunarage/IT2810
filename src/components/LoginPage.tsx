import React, {Component} from 'react';


class LoginPage extends Component<{}, {}> {
    render() {
        return (
            <div className={"login-page"}>
                <h2>Logg inn</h2>
                <form>
                    <div className={"form-group username"}>
                        <label className={"input-label"} htmlFor={"username"}>Brukernavn</label>
                        <input name={"username"} className={"input"} autoFocus placeholder={"OlaNormann"}/>
                    </div>

                    <div className={"form-group password"}>
                        <label className={"input-label"} htmlFor={"password"}>Passord</label>
                        <input name={"password"} className={"input"} type={"password"} autoFocus
                               placeholder={"Passord"}/>
                    </div>
                </form>

            </div>

        );
    }


}

export default LoginPage;