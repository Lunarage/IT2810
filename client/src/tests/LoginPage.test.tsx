import { render, fireEvent } from "@testing-library/react";
import LoginPage from "../components/LoginPage";
import React from "react";
import { Provider } from "react-redux";
import { store } from "../reducers/Reducer";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { toggle_loggedIn } from "../reducers/Actions";

test("LoginPage renders", () => {
    const { container } = render(
        <Provider store={store}>
            <LoginPage />
        </Provider>
    );
    expect(container).toMatchSnapshot();
});

const mockStore = configureStore([]);

describe("My Connected React-Redux Component", () => {
    let store;
    let component;

    beforeEach(() => {
        store = mockStore({
            LoggedIn: false,
            userName: null,
        });

        store.dispatch = jest.fn();

        component = render(
            <Provider store={store}>
                <LoginPage />
            </Provider>
        );
    });
/*
    it("should render with given state from Redux store", () => {
        expect(component.toJSON()).toMatchSnapshot();
    });
*/
    it("should dispatch an action on handle submit", () => {
        renderer.act(() => {
            const button = component.root.findByType("button");
            fireEvent.click(button)
        });

        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(toggle_loggedIn(true));
    });
});
/*
  test('handle login click'), () => {

}
  it('handle home click', () => {
    const { getByRole } = documentBody
    const history = createMemoryHistory()

    fireEvent.click(getByRole('homeButton'));

    expect(history.location.pathname).toBe('/')
  })*/
