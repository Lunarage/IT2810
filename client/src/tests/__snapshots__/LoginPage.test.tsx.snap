// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage renders 1`] = `
<div>
  <div
    class="login-page"
  >
    <div
      class="login-section"
    >
      <h2
        class="login-page-title"
      >
        Log in
      </h2>
      <form
        class="login-page-form"
      >
        <div
          class="form-group username"
        >
          <label
            class="input-label"
            for="username"
          >
            Username
          </label>
          <input
            class="input"
            name="username"
            pattern="[a-zA-Z0-9]{3,}"
            placeholder="OlaNordmann"
            required=""
            title="At least 3 characters: a-z, A-Z, 0-9."
            value=""
          />
        </div>
        <div
          class="form-group password"
        >
          <label
            class="input-label"
            for="password"
          >
            Password
          </label>
          <input
            class="input"
            disabled=""
            name="password"
            placeholder="Password"
            type="password"
          />
        </div>
        <div
          class="form-group"
        />
        <button
          class="login-button"
          type="submit"
        >
          Log in
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<div
  className="login-page"
>
  <div
    className="login-section"
  >
    <h2
      className="login-page-title"
    >
      Log in
    </h2>
    <form
      className="login-page-form"
      onSubmit={[Function]}
    >
      <div
        className="form-group username"
      >
        <label
          className="input-label"
          htmlFor="username"
        >
          Username
        </label>
        <input
          autoFocus={true}
          className="input"
          name="username"
          onChange={[Function]}
          pattern="[a-zA-Z0-9]{3,}"
          placeholder="OlaNordmann"
          required={true}
          title="At least 3 characters: a-z, A-Z, 0-9."
          value=""
        />
      </div>
      <div
        className="form-group password"
      >
        <label
          className="input-label"
          htmlFor="password"
        >
          Password
        </label>
        <input
          className="input"
          disabled={true}
          name="password"
          placeholder="Password"
          type="password"
        />
      </div>
      <div
        className="form-group"
      />
      <button
        className="login-button"
        type="submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;
